### PR TITLE
Add Python to CompositeImplicitAutograd

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #82333

Signed-off-by: Edward Z. Yang <ezyangfb.com>

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -2238,13 +2238,13 @@
...
   dispatch:
-    CPU: eye_out_cpu
+    CPU, Meta: eye_out_cpu
     CUDA: eye_out_cuda
     MPS: eye_out_mps
 
 - func: eye.m_out(int n, int m, *, Tensor(a!) out) -> Tensor(a!)
   dispatch:
-    CPU: eye_out_cpu
+    CPU, Meta: eye_out_cpu
```
Comment (bdhirsh) [inline aten/src/ATen/native/native_functions.yaml:2247]: (wasn't this split out into a separate PR somewhere? np though if you want to land it all together)

File: c10/core/DispatchKeySet.cpp [hunk 2]
```diff
@@ -46,8 +46,8 @@ bool isBackendDispatchKey(DispatchKey t) {
 // math_dispatch_keyset contains all keys in backend_dispatch_keyset and
 // autograd_dispatch_keyset Alias key DispatchKey::CompositeImplicitAutograd
 // maps to [math_dispatch_keyset x full_backend_mask]
-constexpr DispatchKeySet math_dispatch_keyset =
-    backend_dispatch_keyset | autograd_dispatch_keyset;
+constexpr DispatchKeySet math_dispatch_keyset = backend_dispatch_keyset |
+    autograd_dispatch_keyset | DispatchKeySet(DispatchKey::Python);
```
Comment (bdhirsh) [inline c10/core/DispatchKeySet.cpp:50]: This probably deserves a comment for why we're putting the (backend) `Python` key into `math_dispatch_keyset`, and not `backend_dispatch_keyset`?

My understanding is basically: if we did the latter, that would cause all `CompositeExplicitAutograd` kernels to decompose before python subclasses, which... would be pretty BC-breaking at least. Putting it in `math_dispatch_keyset` preserves the current state better, while also fixing some bugs (like preventing re-entrant python dispatch from running `CompositeImplicitAutograd` decomps).

File: test/test_decomp.py [hunk 3]
```diff
@@ -279,6 +279,8 @@ def normalize_op_input_output(f, sample, requires_grad=True):
     # CompositeAutogradImplicit
     # See https://github.com/pytorch/pytorch/issues/81669
     (None, None, "nn.functional.relu6"),
+    (None, None, "nn.functional.mish"),
+    (None, None, "nn.functional.silu"),
```
Comment (bdhirsh) [inline test/test_decomp.py:283]: just curious - what caused these to change?

Comment (ezyang) [reply test/test_decomp.py:283]: They're now getting decomposed, whereas previously we were incorrectly not decomposing them

File: c10/core/DispatchKeySet.cpp (hunk 2 above)
Comment (ezyang) [reply c10/core/DispatchKeySet.cpp:50]: So I actually have no idea. I tried the other version and stuff segfaulted, and so I reverted back to this.

Intuitively, I think the idea is we only want to apply "math" decomps automatically for Python key; the more low level backend explicit autograd decomps should be opt in. But in principle it seems to me that CompositeExplicitAutograd should be ok too, it just isn't.

File: aten/src/ATen/native/native_functions.yaml (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/native/native_functions.yaml:2247]: yes, once I rebase past it this will go away

File: c10/core/DispatchKeySet.cpp (hunk 2 above)
Comment (zou3519) [reply c10/core/DispatchKeySet.cpp:50]: I tried debugging through the PR where we had Python decompose at the CompositeExplicitAutograd level. Some of the segfaults looked like TensorIterator was getting passed a torch_dispatch Tensor Subclass, which is not what we want. So there might just be a mechanism problem somewhere

File: test/test_functionalization.py [hunk 4]
```diff
@@ -116,10 +116,21 @@ def f(x):
 
 
 def forward(self, a_1):
+<<<<<<< HEAD
```
Comment (zou3519) [inline test/test_functionalization.py:119]: You probably want to fix the merge conflict?

**Review by zou3519 — APPROVED** [review]
LGTM modulo the source control markers from a merge conflict

**Comment by tarun292** [comment]
@ezyang the changes here seem to be breaking some of the executorch tests. Specifically the graphs trace before and after this change seem to be different causing a bunch of test failures. Following up on the internal diff with more details.

**Comment by mikaylagawarecki** [comment]
Hi! Just want to double check for release notes purposes that this bc-breaking changewas never relanded (seems like https://github.com/pytorch/pytorch/pull/82471 was closed), is that correct? Thanks!

**Comment by ezyang** [comment]
Yes we abandoned it
